### PR TITLE
[ML] validate model definition on start deployment

### PR DESCRIPTION
**Pull request by benwtrent**
When a deployment is started, we do not validate that the definition documents are all present and not truncated.

This commit adds a validation on _start that prevents a bad state from occurring where the deployment starts, but the model is incorrectly defined, or some unknown error occurs to late in the deployment process.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java [hunk 1]
```diff
@@ -73,7 +71,7 @@ public Request(String modelId, BytesReference definition, int part, long totalDe
         public Request(StreamInput in) throws IOException {
             super(in);
             this.modelId = in.readString();
-            this.definition = in.readBytesReference();
+            this.definition = in.readString();
```
Comment (przemekwitek) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:74]: Can the wire type of the field be changed like this between versions?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java [hunk 2]
```diff
@@ -270,6 +282,74 @@ private void deleteFailedDeployment(
...
+                        listener.onFailure(
+                            ExceptionsHelper.badRequestException(
+                                "[{}] model definition is missing required fields {}, unable to be deployed",
+                                modelId,
+                                List.of(requiredSourceFields)
+                            )
+                        );
+                        return;
+                    }
+                    Set<String> diff = Sets.difference(fields.keySet(), Set.of(requiredSourceFields));
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java:329]: nit: `Set.of(requiredSourceFields)` can be lifted out of the for loop so it is created everytime

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java [hunk 3]
```diff
@@ -270,6 +282,74 @@ private void deleteFailedDeployment(
...
+                QueryBuilders.constantScoreQuery(
+                    QueryBuilders.boolQuery()
+                        .filter(QueryBuilders.termQuery(TrainedModelConfig.MODEL_ID.getPreferredName(), modelId))
+                        .filter(
+                            QueryBuilders.termQuery(InferenceIndexConstants.DOC_TYPE.getPreferredName(), TrainedModelDefinitionDoc.NAME)
+                        )
+                )
+            )
+            .setFetchSource(requiredSourceFields, new String[0])
+            .setSize(MAX_NUM_DEFINITION_DOCS)
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java:307]: What if there are > 100 definition docs? The summed sizes will never equal expected total.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java [hunk 4]
```diff
@@ -57,12 +55,12 @@ public static Request parseRequest(String modelId, int part, XContentParser pars
         }
 
         private final String modelId;
-        private final BytesReference definition;
+        private final String definition;
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:58]: The reason for making this change is not clear to me. We expect the definition to be base64 encoded, Java UTF-16 characters require 2 bytes to store each base64 character whereas the raw bytes (given UFT-8 encoding) will use half that space.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java (hunk 1 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:74]: this is going in v8. This was only ever put in v8

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelDefinitionPartAction.java [hunk 5]
```diff
@@ -6,6 +6,8 @@
  */
 package org.elasticsearch.xpack.ml.action;
 
+import com.unboundid.util.Base64;
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelDefinitionPartAction.java:9]: Auto import!!! ARG!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java (hunk 4 above)
Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:58]: Reason for the change is to do `definition_length` we need to decode the base64 values. To do that, it needs to be a string. So, might as well keep it as a string

Comment (benwtrent) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/PutTrainedModelDefinitionPartAction.java:58]: It isn't a big deal, I can switch it back, I will just have to transform it to a string when decoding it.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelDefinitionPartAction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutTrainedModelDefinitionPartAction.java:9]: Ah, and the java util one has a `byte[] -> byte[]`. Brilliant!

**Review by davidkyle — APPROVED** [review]
LGTM

**Comment by davidkyle** [comment]
🤔 💭 I think it is also worth checking that every doc has the same value for `total_definition_length` at start.

I can do that in a follow up PR

**Pull request merged**